components/builder-scheduler/src/data_store.rs

```diff
@@ -12,215 +12,335 @@
+                                $$
+                                DECLARE
+                                    r groups % rowtype;
+                                BEGIN
+                                    FOR r IN
+                                        SELECT * FROM groups
+                                        WHERE group_state = 'Pending'
+                                        ORDER BY created_at ASC
+                                        FOR UPDATE SKIP LOCKED
+                                        LIMIT $1
+                                    LOOP
+                                        UPDATE groups SET group_state='Dispatching', updated_at=now() WHERE id=r.id RETURNING * INTO r;
+                                        RETURN NEXT r;
+                                    END LOOP;
+                                  RETURN;
+                                END
+                                $$ LANGUAGE plpgsql VOLATILE"#)?;
+
+        // Update the state of a group
+        migrator.migrate("scheduler",
+                         r#"CREATE OR REPLACE FUNCTION set_group_state_v1 (gid bigint, gstate text) RETURNS void AS $$
+                            BEGIN
+                                UPDATE groups SET group_state=gstate, updated_at=now() WHERE id=gid;
+                            END
+                         $$ LANGUAGE plpgsql VOLATILE"#)?;
+
+        // Update the state of a project
+        migrator.migrate("scheduler",
+                          r#"CREATE OR REPLACE FUNCTION set_project_state_v1 (pid bigint, jid bigint, state text) RETURNS void AS $$
+                             BEGIN
+                                 UPDATE projects SET project_state=state, job_id=jid, updated_at=now() WHERE id=pid;
+                             END
+                          $$ LANGUAGE plpgsql VOLATILE"#)?;
+
+        migrator.migrate("scheduler",
+                         r#"CREATE INDEX pending_groups_index_v1 on groups(created_at) WHERE group_state = 'Pending'"#)?;
+
+        // Retrieve a group project
+        migrator.migrate("scheduler",
+                  r#"CREATE OR REPLACE FUNCTION find_project_v1 (gid bigint, name text) RETURNS SETOF projects AS $$
+                         BEGIN
+                           RETURN QUERY SELECT * FROM projects WHERE owner_id = gid AND project_name = name;
+                           RETURN;
+                         END
+                         $$ LANGUAGE plpgsql STABLE"#)?;
+
         Ok(())
     }
 
-    pub fn create_group(&mut self, project_names: Vec<String>) -> Result<Group> {
-        let mut projects = Vec::new();
-        for name in project_names {
-            let project = ProjectData {
-                name: name,
-                state: proto::ProjectState::NotStarted,
-            };
-            projects.push(project);
-        }
+    pub fn create_group(&self, project_names: Vec<String>) -> Result<Group> {
```

Comment:
Rather than take a `Vec<String>` here, I would change this to take the incoming protocol protobuf. That's how the `originsrv` works now, and it's been much nicer.